web/handlers.go

```diff
@@ -190,13 +191,26 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if len(token) != 0 {
 		session, err := c.App.GetSession(token)
 		if err != nil {
-			c.Log.Info("Invalid session", mlog.Err(err))
-			if err.StatusCode == http.StatusInternalServerError {
-				c.Err = err
-			} else if h.RequireSession {
-				c.RemoveSessionCookie(w, r)
-				c.Err = model.NewAppError("ServeHTTP", "api.context.session_expired.app_error", nil, "token="+token, http.StatusUnauthorized)
+			if c.App.Srv().License() != nil && *c.App.Srv().License().Features.Cloud && tokenLocation == app.TokenLocationCloudHeader {
+				// Check to see if this provided token matches our CWS Token
+				session, err = c.App.GetCloudSession(token)
+				if err != nil {
+					c.Log.Info("Invalid CWS token", mlog.Err(err))
+					c.Err = err
+				} else {
+					c.App.SetSession(session)
+				}
+
+			} else {
+				c.Log.Info("Invalid session", mlog.Err(err))
+				if err.StatusCode == http.StatusInternalServerError {
+					c.Err = err
+				} else if h.RequireSession {
+					c.RemoveSessionCookie(w, r)
+					c.Err = model.NewAppError("ServeHTTP", "api.context.session_expired.app_error", nil, "token="+token, http.StatusUnauthorized)
+				}
 			}
+
 		} else if !session.IsOAuth && tokenLocation == app.TokenLocationQueryString {
 			c.Err = model.NewAppError("ServeHTTP", "api.context.token_provided.app_error", nil, "token="+token, http.StatusUnauthorized)
 		} else {
```

Comment:
Given that we don't have session in this case, probably we don't need the CSRFToken or handle in some other way

Reply:
This is no longer applicable as of the latest diff. Going to resolve this one and we can discuss here: https://github.com/mattermost/mattermost-server/pull/16319#discussion_r527795317